### PR TITLE
Fix erroneous prefix matches in HFSX binary comparison

### PR DESCRIPTION
**Pull request by DHowett**
This fixes the case where dir_205123 would erroneously match dir_205.

This commit completes the FIXME and removes (obviates) the special cases
for zero-length keys on both sides.

Signed-off-by: Dustin L. Howett <dustin@howett.net>

This is perhaps a more contentious change; it ditches the memcmp in favor of a hand-rolled implementation. The early returns for zero-length keys aren't necessary this way, and it does seem to address the FIXME.

### CODE REVIEW

File: lib/libhfs/libhfs.c [hunk 1]
```diff
@@ -2484,25 +2484,31 @@ hfslib_compare_catalog_keys_bc (
...
-		if(((const hfs_catalog_key_t*)a)->name.length == 0)
-			return -1;
-		if(((const hfs_catalog_key_t*)b)->name.length == 0)
-			return 1;
+			if (ac == bc)
+				continue;
 
-		/* FIXME: This does a byte-per-byte comparison, whereas the HFS spec
-		 * mandates a uint16_t chunk comparison. */
-		return memcmp(((const hfs_catalog_key_t*)a)->name.unicode,
```
Comment (0x09) [inline lib/libhfs/libhfs.c]: I see the issue. It looks like you could keep most of the existing function including the memcmp, but capture its result and return the difference of the lengths if nonzero. I'd prefer to keep the diffs against the NetBSD code small where possible to facilitate keeping them in sync, but also memcmp is likely to be vectorized and should be more performant here.

**Comment by DHowett** [comment]
Regarding keeping the diffs from NetBSD small: that makes sense! I've amended the changeset. :smile: Thanks!

**Comment by 0x09** [comment]
Just a heads up, I merged in the current top of tree NetBSD code just now at 0x09/hfsfuse@7d70266  (hfsfuse was a few versions behind, but I put it off since the changes are entirely cosmetic). Since it's a large reformatting, there will likely be conflicts if you have edited anything else from lib/libhfs locally. But it should be easier to compare the two / apply any changes back.

**Comment by DHowett** [comment]
Excellent. Thanks!

**Pull request merged**